cs/src/core/ClientSession/FASTERAsync.cs

```diff
@@ -0,0 +1,365 @@
+        {
+            if (async)
+                clientSession.UnsafeResumeThread();
+
+            var previousState = SystemState.Make(ctx.phase, ctx.version);
+            var finalState = SystemState.Copy(ref _systemState);
+
+            while (finalState.phase == Phase.INTERMEDIATE)
+            {
+                finalState = SystemState.Copy(ref _systemState);
+            }
+
+            // We need to move from previousState to finalState one step at a time
+            do
+            {
+                var currentState = default(SystemState);
+                if (previousState.word == finalState.word)
+                {
+                    currentState.word = previousState.word;
+                }
+                else if (previousState.version < finalState.version)
+                {
+                    if (finalState.phase <= Phase.PREPARE)
+                        previousState.version = finalState.version;
+                    else
+                        previousState.version = finalState.version - 1;
+                    currentState = GetNextState(previousState, CheckpointType.FULL);
+                }
+                else
+                {
+                    currentState = GetNextState(previousState, _checkpointType);
+                }
+
+                switch (currentState.phase)
+                {
+                    case Phase.PREP_INDEX_CHECKPOINT:
+                        {
+                            if (!ctx.markers[EpochPhaseIdx.PrepareForIndexCheckpt])
+                            {
+                                ctx.markers[EpochPhaseIdx.PrepareForIndexCheckpt] = true;
+                            }
+                            if (epoch.MarkAndCheckIsComplete(EpochPhaseIdx.PrepareForIndexCheckpt, ctx.version))
+                            {
+                                GlobalMoveToNextCheckpointState(currentState);
+                            }
+                            break;
+                        }
+                    case Phase.INDEX_CHECKPOINT:
+                        {
+                            if (_checkpointType == CheckpointType.INDEX_ONLY)
+                            {
+                                // Reseting the marker for a potential FULL or INDEX_ONLY checkpoint in the future
+                                ctx.markers[EpochPhaseIdx.PrepareForIndexCheckpt] = false;
+                            }
+
+                            if (async && !IsIndexFuzzyCheckpointCompleted())
+                            {
+                                clientSession.UnsafeSuspendThread();
+                                await IsIndexFuzzyCheckpointCompletedAsync();
+                                clientSession.UnsafeResumeThread();
+                            }
+                            GlobalMoveToNextCheckpointState(currentState);
```

Comment:
Don't we need to check if `IsIndexFuzzyCheckpointCompleted()` for the non-async case as well?